interface-definitions/include/firewall/gre.xml.i

```diff
@@ -0,0 +1,126 @@
+    <help>GRE fields to match</help>
+  </properties>
+  <children>
+    <node name="flags">
+      <properties>
+        <help>GRE flag bits to match</help>
+      </properties>
+      <children>
+        <node name="key">
+          <properties>
+            <help>Header includes optional key field</help>
+          </properties>
+          <children>
+            <leafNode name="unset">
+              <properties>
+                <help>Header does not include optional key field</help>
+                <valueless/>
+              </properties>
+            </leafNode>
+          </children>
+        </node>
+        <node name="checksum">
+          <properties>
+            <help>Header includes optional checksum</help>
+          </properties>
+          <children>
+            <leafNode name="unset">
+              <properties>
+                <help>Header does not include optional checksum</help>
+                <valueless/>
+              </properties>
+            </leafNode>
+          </children>            
+        </node>
+        <node name="sequence">
+          <properties>
+            <help>Header includes a sequence number field</help>
+          </properties>
+          <children>
+            <leafNode name="unset">
+              <properties>
+                <help>Header does not include a sequence number field</help>
+                <valueless/>
+              </properties>
+            </leafNode>
+          </children>
+        </node>
+      </children>
+    </node>
+    <leafNode name="inner-proto">
+      <properties>
+        <help>EtherType of encapsulated packet</help>
+        <completionHelp>
+          <list>ip ip6 arp 802.1q 802.1ad</list>
+        </completionHelp>
+        <valueHelp>
+          <format>u32:0-65535</format>
+          <description>Ethernet protocol number</description>
+        </valueHelp>
+        <valueHelp>
+          <format>u32:0x0-0xFFFF</format>
+          <description>Ethernet protocol number (hex)</description>
+        </valueHelp>
+        <valueHelp>
+          <format>ip</format>
+          <description>IPv4</description>
+        </valueHelp>
+        <valueHelp>
+          <format>ip6</format>
+          <description>IPv6</description>
+        </valueHelp>
+        <valueHelp>
+          <format>arp</format>
+          <description>Address Resolution Protocol</description>
+        </valueHelp>
+        <valueHelp>
+          <format>802.1q</format>
+          <description>VLAN-tagged frames (IEEE 802.1q)</description>
+        </valueHelp>
+        <valueHelp>
+          <format>802.1ad</format>
+          <description>Provider Bridging (IEEE 802.1ad, Q-in-Q)</description>
+        </valueHelp>
+        <valueHelp>
+          <format>gretap</format>
+          <description>Transparent Ethernet Bridging (L2 Ethernet over GRE, gretap)</description>
+        </valueHelp>
+        <constraint>
+          <regex>(ip|ip6|arp|802.1q|802.1ad|gretap|\d+|0x[0-9a-fA-F]+)</regex>
```

Comment:
Please drop `\d+` from the regex and rather use this instead, as validators are logically OR'ed:

```

         <constraint>
-          <regex>(ip|ip6|arp|802.1q|802.1ad|gretap|\d+|0x[0-9a-fA-F]+)</regex>
+          <regex>(ip|ip6|arp|802.1q|802.1ad|gretap|0x[0-9a-fA-F]+)</regex>
+          <validator name="numeric" argument="--range 0-65535"/>
         </constraint>
```

Please also not that the hex validation logic is NOT precise enough:

```
cpo@LR2.wue3# set firewall ipv4 name foo rule 10 gre inner-proto 0x20000
```

This works and is an illegal value. I think it should be: `0x[0-9a-fA-F]{1,4}` instead.

@dmbaturin we probably should also make the numeric validator HEX aware.

Reply:
> as validators are logically OR'ed
Thanks. I was trying to figure out a way to do that neatly. There is a range check in verify_rule() but it's nicer to have it driven from the CLI with immediate feedback to the user. 

> make the numeric validator HEX aware
dmbaturin picked that up already from earlier comments: https://vyos.dev/T6622. It would be the ideal way to do this - the regex stuff is just to get something workable until a radix-tolerant validator is available and can be retrofitted in.